src/intel/m128_.rs

```diff
@@ -0,0 +1,264 @@
+#![allow(clippy::transmute_ptr_to_ptr)]
+
+//! This module is for the `m128` wrapper type, its bonus methods, and all
+//! necessary trait impls.
+//!
+//! Intrinsics should _not_ be in this module! They should all be free-functions
+//! in the other modules, sorted by CPU target feature.
+
+use super::*;
+
+/// The data for a 128-bit SSE lane.
+///
+/// * This is _very similar to_ having `[f32; 4]`. The main difference is that
+///   it's aligned to 16 instead of just 4, and of course you can perform
+///   various intrinsic operations on it.
+/// * You can use `as_ref` and `as_mut` to view the type as if it was an array,
+///   and from there you _could_ access an individual lane via indexing if you
+///   wanted. However, doing this will usually kill your performance if you're
+///   in the middle of a series of operations. The CPU has to move the type out
+///   of register and into memory, then index the memory. In other words, you
+///   should index the individual lanes as little as possible. Accordingly, we
+///   make you use a "more obvious" trait if you want to do it.
+#[repr(transparent)]
+#[allow(non_camel_case_types)]
+pub struct m128(pub __m128);
+
+#[test]
+fn test_m128_size_align() {
+  assert_eq!(core::mem::size_of::<m128>(), 16);
+  assert_eq!(core::mem::align_of::<m128>(), 16);
+}
+
+impl m128 {
+  /// Transmutes the `m128` to an array.
+  ///
+  /// Same as `m.into()`, just lets you be more explicit about what's happening.
+  #[must_use]
+  #[inline(always)]
+  pub fn to_array(self) -> [f32; 4] {
+    self.into()
+  }
+
+  /// Transmutes an array into `m128`.
+  ///
+  /// Same as `m128::from(arr)`, it just lets you be more explicit about what's
+  /// happening.
+  #[must_use]
+  #[inline(always)]
+  pub fn from_array(f: [f32; 4]) -> Self {
+    f.into()
+  }
+}
+
+#[cfg(feature = "bytemuck")]
+unsafe impl bytemuck::Zeroed for m128 {}
+#[cfg(feature = "bytemuck")]
+unsafe impl bytemuck::Pod for m128 {}
+
+impl AsRef<[f32; 4]> for m128 {
+  #[must_use]
+  #[inline(always)]
+  fn as_ref(&self) -> &[f32; 4] {
+    unsafe { core::mem::transmute(self) }
+  }
+}
+
+impl AsMut<[f32; 4]> for m128 {
+  #[must_use]
+  #[inline(always)]
+  fn as_mut(&mut self) -> &mut [f32; 4] {
+    unsafe { core::mem::transmute(self) }
+  }
+}
+
+impl Clone for m128 {
+  #[must_use]
+  #[inline(always)]
+  fn clone(&self) -> Self {
+    *self
+  }
+}
+impl Copy for m128 {}
+
+impl Default for m128 {
+  #[must_use]
+  #[inline(always)]
+  fn default() -> Self {
+    unsafe { core::mem::zeroed() }
+  }
+}
+
+impl From<[f32; 4]> for m128 {
+  #[must_use]
+  #[inline(always)]
+  fn from(arr: [f32; 4]) -> Self {
+    unsafe { core::mem::transmute(arr) }
```

Comment:
It's not because the value passes by value, so it gets moved from an align4 input location to an align16 output location.

If this was a reference transmute we'd need to be a lot more careful, but for by-value transmutes it's safe.